torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -454,6 +454,7 @@ bool ProcessGroupNCCL::WorkNCCL::checkTimeout(
       " milliseconds before timing out.");
 
   LOG(ERROR) << exceptionMsg;
+  std::this_thread::sleep_for(std::chrono::milliseconds(10 * kWatchdogThreadSleepMillis));
```

Comment:
How can we let user to tune this time period? Or we believe 10 secs are enough?

Reply:
The exact tuning is a bit tricky. The thinking is to give enough time for other ranks to log their watchdog errors, before this process terminates and causes torchX to tear down the rest of the processes. Having a multiple of the kWatchdogSleepMillis seems reasonable because that is the interval other processes / nodes will run their watchdog, so depending on clock skew they should be able to report their errors.

Reply:
I agree with the direction we want to go here but I guess my question is more around whether this is indeed helping address the painpoint or not, maybe we can ask cutomers to test it out on LLM training?

Reply:
@fduwjj on my local training runs, I confirmed all processes report their watchdog errors with this change whereas before only one would report. but yes, we can do more comprehensive testing.

Reply:
I see.

Reply:
What additional info do you get with this extra 10 seconds?

Shouldn't this be done by the process coordinator instead?

Reply:
@kumpera Agree it should be done ideally by the process coordinator (torchX / elastic). We should align with the elastic folks here to figure out a sustainable solution.

In the meantime, the value add of this stopgap is that it lets all processes that are going to crash with watchdog have a higher chance of reporting that this is the error. For internal logging systems, this allows much better debugability and error attribution. I confirmed this on a few runs where I simulated errors and ensured all ranks that fire watchdog have that logged. If the value of that is sufficient we can probably ship this as a stopgap

Reply:
Is the motivation here to make all ranks fail with the same root cause error so that the user can take a look at the logs for **any** rank and see (similar) errors? If so, internally at Meta this shouldn't be an issue (even without this change) since the way torchelastic's error handler/reporter works internally conforms to the expectation of the internal schedulers (e.g. Mast). Each node's elastic agent will only propagate the "first" (chronologically) error. The scheduler then further propagates the first error (out of the firsts) to the end user (and into scuba tables).

To see ALL the exceptions/errors/signals for every rank, check out the torchelastic error scuba table (if my memory serves right its called "PET errors" - we never renamed the scuba table to torchelastic)

Reply:
Also FWIW, happy to jump on VC to discuss long term solutions here with torchx/elastic (cc @d4l3k , @drdarshan)